### PR TITLE
MM-31322: redirect permalink handler

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Some time ago, support for `_redirect` was added to the webapp permalink handler code. The exact semantics there effectively replaced it with the current team name, allowing the server to generate permalinks to DMs and GMs without having to know more context about the enduser and which teams they access.

This PR extends the mobile application to do same, though it works even better here, since it resolves non-DM and non-GM permalinks across team boundaries. We may revisit the webapp implementation to do same at some point, but all of this is only lightly used at present.

This is my first stab at writing a Detox test, though the Cypress familiarity made it relatively straightforward. Big thanks to @josephbaylon for various assists :)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-31322

#### Checklist
- [x] Added or updated unit tests (required for all new features)

#### Device Information
This PR was tested on: iOS Simulator

### CODE REVIEW

File: app/actions/views/permalink.ts [hunk 1]
```diff
@@ -15,31 +15,37 @@ export let showingPermalink = false;
...
-
-                const options = {
-                    layout: {
-                        componentBackgroundColor: changeOpacity('#000', 0.2),
-                    },
-                };
-
-                showingPermalink = true;
-                showModalOverCurrentContext(screen, passProps, options);
+        // For the special, "_redirect" path, just use the currently selected team.
```
Comment (lieut-data) [inline app/actions/views/permalink.ts:18]: This is just to avoid the error of trying to load an invalid team.

File: app/components/markdown/markdown_link/markdown_link.js [hunk 2]
```diff
@@ -61,7 +62,11 @@ export default class MarkdownLink extends PureComponent {
                 const {intl} = this.context;
                 this.props.actions.handleSelectChannelByName(match.channelName, match.teamName, errorBadChannel.bind(null, intl));
             } else if (match.type === DeepLinkTypes.PERMALINK) {
-                onPermalinkPress(match.postId, match.teamName);
+                if (match.teamName === PERMALINK_GENERIC_TEAM_NAME_REDIRECT) {
```
Comment (lieut-data) [inline app/components/markdown/markdown_link/markdown_link.js:65]: Here's the "magic".

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        // * Verify permalink post list has the expected target message
+        await PermalinkScreen.toBeVisible();
+        const {postListPostItem: permalinkPostItem} = await PermalinkScreen.getPostListPostItem(permalinkTargetPost.id, permalinkTargetPost.message);
+        await waitFor(permalinkPostItem).toBeVisible();
+
+        // * Dismiss the permalink screen by jumping to recent messages
+        await PermalinkScreen.jumpToRecentMessages();
+    };
+
+    it('[tm4j_id]_[step] should support _redirect to public channel post', async () => {
```
Comment (lieut-data) [inline detox/e2e/test/messaging/permalink.e2e.js:58]: I'm not terribly familiar with creating TM4J tasks -- could use some help here!

**Review by lieut-data — COMMENTED** [review]
As discussed, after rebasing against `e2e-channels-list`, my tests started failing with:
```
    Test Failed: View “<RCTView: 0x7fd8e152ad70>” is not visible: view does not pass visibility threshold (0% visible of 75% required)
    TIP: To print view hierarchy on failed actions/matches, use log-level verbose or higher.

      142 |     openMainSidebar = async () => {
      143 |         // # Open main sidebar
    > 144 |         await this.mainSidebarDrawerButton.tap();
          |                                            ^
      145 |         await expect(MainSidebar.channelsList).toBeVisible();
      146 |         await MainSidebar.toBeVisible();
      147 |     }
```

Previously, the non-DM test was passing. Though perhaps I've just messed up something else?

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        // # Open channel screen
+        await ChannelScreen.open(user);
+    });
+
+    afterAll(async () => {
+        await ChannelScreen.logout();
+    });
+
+    const expectPermalinkTargetMessage = async (permalinkTargetPost) => {
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:33]:
```suggestion
    const expectPermalinkTargetMessage = async (permalinkTargetPost, permalinkTargetChannel) => {
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        // # Post a message in the test channel referencing the given permalink.
+        const permalinkMessage = `[permalink](/_redirect/pl/${permalinkTargetPost.id})`;
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:35]:
```suggestion
        const permalinkLabel = `permalink-${Date.now().toString()}`;
        const permalinkMessage = `[${permalinkLabel}](/_redirect/pl/${permalinkTargetPost.id})`;
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        // # Tap the channel permalink
+        element(by.text('permalink')).tap({x: 5, y: 10});
+
+        // * Verify permalink post list has the expected target message
+        await PermalinkScreen.toBeVisible();
+        const {postListPostItem: permalinkPostItem} = await PermalinkScreen.getPostListPostItem(permalinkTargetPost.id, permalinkTargetPost.message);
+        await waitFor(permalinkPostItem).toBeVisible();
+
+        // * Dismiss the permalink screen by jumping to recent messages
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:54]:
```suggestion
        // # Dismiss the permalink screen by jumping to recent messages
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        await PermalinkScreen.jumpToRecentMessages();
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:55]:
```suggestion
        await PermalinkScreen.jumpToRecentMessages();
        
        // * Verify user is on channel where message is posted
        await expect(ChannelScreen.channelNavBarTitle).toHaveText(permalinkTargetChannel.display_name);
        const {postListPostItem: channelPostItem} = await ChannelScreen.getPostListPostItem(permalinkTargetPost.id, permalinkTargetPost.message);
        await expect(channelPostItem).toBeVisible();
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    it('[tm4j_id]_[step] should support _redirect to public channel post', async () => {
+        // # Post a test message in a public channel
+        const permalinkTargetMessage = 'post in Town Square';
+        const permalinkTargetPost = await Post.apiCreatePost({
+            channelId: townSquareChannel.id,
+            message: permalinkTargetMessage,
+        });
+
+        expectPermalinkTargetMessage(permalinkTargetPost.post);
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:66]:
```suggestion
        await expectPermalinkTargetMessage(permalinkTargetPost.post, townSquareChannel);
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        // # Post a test message in a DM
+        const {user: dmOtherUser} = await User.apiCreateUser({prefix: 'testchannel-1'});
+        const {channel: directMessageChannel} = await Channel.apiCreateDirectChannel([testUser.id, dmOtherUser.id]);
+        const permalinkTargetMessage = 'post in DM';
+        const permalinkTargetPost = await Post.apiCreatePost({
+            channelId: directMessageChannel.id,
+            message: permalinkTargetMessage,
+        });
+
+        expectPermalinkTargetMessage(permalinkTargetPost.post);
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:79]:
```suggestion
        await expectPermalinkTargetMessage(permalinkTargetPost.post, directMessageChannel);
```

**Review by josephbaylon — CHANGES_REQUESTED** [review]
Here are the changes I had locally to make it work. Also, regarding tm4j keys, I'll create tm4j test cases later if none exist and provide the keys to you, unless you know the tm4j test cases this PR is based from.

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            message: permalinkMessage,
+        });
+
+        // # Go to test channel
+        await ChannelScreen.openMainSidebar();
+        await MainSidebar.getChannelByDisplayName(testChannel.display_name).tap();
+        await ChannelScreen.toBeVisible();
+
+        // # Tap the channel permalink
+        element(by.text('permalink')).tap({x: 5, y: 10});
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:47]:
```suggestion
        element(by.text(permalinkLabel)).tap({x: 5, y: 10});
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+        // * Verify permalink post list has the expected target message
+        await PermalinkScreen.toBeVisible();
+        const {postListPostItem: permalinkPostItem} = await PermalinkScreen.getPostListPostItem(permalinkTargetPost.id, permalinkTargetPost.message);
+        await waitFor(permalinkPostItem).toBeVisible();
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:52]:
```suggestion
        await expect(permalinkPostItem).toBeVisible();
```

File: detox/e2e/test/messaging/permalink.e2e.js (hunk 8 above)
Comment (lieut-data) [reply detox/e2e/test/messaging/permalink.e2e.js:66]: Ah, this was the key difference that broke things. I'm so used to Cypress' pseudo-promise semantics that I complete forgot to actually `await` here. Thank you!

File: detox/e2e/test/messaging/permalink.e2e.js (hunk 7 above)
Comment (lieut-data) [reply detox/e2e/test/messaging/permalink.e2e.js:55]: Ooh, good addition!

File: detox/e2e/test/messaging/permalink.e2e.js (hunk 5 above)
Comment (lieut-data) [reply detox/e2e/test/messaging/permalink.e2e.js:35]: Thank you for this -- it wasn't obvious until the tests were running together why this wouldn't work as well.

**Comment by lieut-data** [comment]
Really appreciate the help, @josephbaylon!

> Also, regarding tm4j keys, I'll create tm4j test cases later if none exist and provide the keys to you, unless you know the tm4j test cases this PR is based from.

None exist at present, as this is all "new" functionality. Happy to update once available :)

File: app/actions/views/permalink.ts (hunk 1 above)
Comment (migbot) [reply app/actions/views/permalink.ts:18]: How does the string `_redirect` end up here? Doesn't the current team name get passed to `showPermalink` when `match.teamName` is `_redirect`?

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        // # Dismiss the permalink screen by jumping to recent messages
+        await PermalinkScreen.jumpToRecentMessages();
+
+        // * Verify user is on channel where message is posted
+        await expect(ChannelScreen.channelNavBarTitle).toHaveText(permalinkTargetChannel.display_name);
+        const {postListPostItem: channelPostItem} = await ChannelScreen.getPostListPostItem(permalinkTargetPost.id, permalinkTargetPost.message);
+        await expect(channelPostItem).toBeVisible();
+    };
+
+    it('[tm4j_id]_[step] should support _redirect to public channel post', async () => {
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:64]:
```suggestion
    it('MM-T3805_1 should support _redirect to public channel post', async () => {
```

File: detox/e2e/test/messaging/permalink.e2e.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        const permalinkTargetMessage = 'post in Town Square';
+        const permalinkTargetPost = await Post.apiCreatePost({
+            channelId: townSquareChannel.id,
+            message: permalinkTargetMessage,
+        });
+
+        await expectPermalinkTargetMessage(permalinkTargetPost.post, townSquareChannel);
+    });
+
+    it('[tm4j_id]_[step] should support _redirect to DM post', async () => {
```
Comment (josephbaylon) [inline detox/e2e/test/messaging/permalink.e2e.js:75]:
```suggestion
    it('MM-T3805_2 should support _redirect to DM post', async () => {
```

**Review by josephbaylon — CHANGES_REQUESTED** [review]
Created a new TM4J case [MM-T3805](https://mattermost.atlassian.net/projects/MM?selectedItem=com.atlassian.plugins.atlassian-connect-plugin%3Acom.kanoah.test-manager__main-project-page#!/testCase/MM-T3805)

File: app/actions/views/permalink.ts (hunk 1 above)
Comment (lieut-data) [reply app/actions/views/permalink.ts:18]: Hmm, I thought I had purged the vestigial parts of my initial investigation, but it looks like you're right: this isn't needed. Early on, I was encountering issues with what I thought was a proactive metadata fetch from orthogonal code, but I can no longer reproduce same. I've reverted all the corresponding changes.

**Comment by lieut-data** [comment]
@migbot / @josephbaylon, I'm also wondering about is handling of "deep links". What's the easiest way to "make one" so I can test that these are routed correctly?

**Comment by josephbaylon** [comment]
> @migbot / @josephbaylon, I'm also wondering about is handling of "deep links". What's the easiest way to "make one" so I can test that these are routed correctly?

I'm not sure if this is what you're looking for but in your device you can copy the link of the post, then in one of the device browsers, paste and modify such that,
https://hostname/core/pl/123 --> mattermost://hostname/_redirect/pl/123

Without redirect mattermost://hostname/core/pl/123 should open in the MM app. I'm guessing you want to test if redirect link also opens properly.

**Comment by josephbaylon** [comment]
In terms of e2e, you don't need to add it. Detox has current limitation right now to test two apps easily (in this case, browser and MM)

**Comment by lieut-data** [comment]
All right, lots of learning on my part -- thanks for the patience! I've added support for deeplinks using the same heuristics as markdown.

**Review by josephbaylon — APPROVED** [review]
LGTM. Redirect permalinks now work on both regular app markup and deeplinks. Also smoke tested regular permalinks.

Tested on both iPhone X (14.3) and Pixel 3 XL (11) devices.

Also reviewed and tested detox e2e.

**Review by migbot — APPROVED** [review]
LGTM!

File: app/components/markdown/markdown_link/index.js [hunk 14]
```diff
@@ -5,14 +5,18 @@ import {bindActionCreators} from 'redux';
...
 
 import MarkdownLink from './markdown_link';
 
 function mapStateToProps(state) {
+    const team = getCurrentTeam(state);
+
     return {
         serverURL: getCurrentUrl(state),
         siteURL: getConfig(state).SiteURL,
+        currentTeamName: team && team.name,
```
Comment (avinashlng1080) [inline app/components/markdown/markdown_link/index.js:19]: I think this is also equivalent 

`currentTeamName: getCurrentTeam(state)?.name,`

File: app/components/post_list/index.js [hunk 15]
```diff
@@ -23,13 +24,16 @@ function makeMapStateToProps() {
...
+        const team = getCurrentTeam(state);
+
         return {
             deepLinkURL: state.views.root.deepLinkURL,
             postIds,
             initialIndex,
             serverURL: getCurrentUrl(state),
             siteURL: getConfig(state).SiteURL,
             theme: getTheme(state),
+            currentTeamName: team && team.name,
```
Comment (avinashlng1080) [inline app/components/post_list/index.js:36]: Same as above comment.

File: app/components/post_list/index.js [hunk 16]
```diff
@@ -23,13 +24,16 @@ function makeMapStateToProps() {
             initialIndex = postIds.indexOf(ownProps.highlightPostId);
         }
 
+        const team = getCurrentTeam(state);
+
```
Comment (lieut-data) [inline app/components/post_list/index.js:28]:
```suggestion
```

File: app/components/post_list/index.js (hunk 15 above)
Comment (lieut-data) [reply app/components/post_list/index.js:36]:
```suggestion
            getCurrentTeam(state)?.name,
```

**Comment by lieut-data** [comment]
Thanks, @avinashlng1080! I've applied your excellent suggestions :)

**Comment by avinashlng1080** [comment]
Thanks @lieut-data .

Once the CI error is corrected, I will approve the PR.

The error is : 

```
/home/circleci/mattermost-mobile/app/components/post_list/index.js
  34:33  error  Parsing error: '{' expected
```

File: app/components/post_list/index.js [hunk 17]
```diff
@@ -24,16 +24,14 @@ function makeMapStateToProps() {
...
-
         return {
             deepLinkURL: state.views.root.deepLinkURL,
             postIds,
             initialIndex,
             serverURL: getCurrentUrl(state),
             siteURL: getConfig(state).SiteURL,
             theme: getTheme(state),
-            currentTeamName: team && team.name,
+            getCurrentTeam(state)?.name,
```
Comment (lieut-data) [inline app/components/post_list/index.js:34]:
```suggestion
            currentTeamName: getCurrentTeam(state)?.name,
```

**Comment by lieut-data** [comment]
Oops! I guess I should have just done the changes from the CLI instead of hacking GitHub's suggestion mode :p -- should be fixed now.

(I'm surprised I can merge without the CI passing? Is that intentional)?

**Review by avinashlng1080 — APPROVED** [review]
LGTM

**Comment by avinashlng1080** [comment]
@lieut-data I'm not sure if this is intentional.  @migbot do you know if it is intentional ?

**Comment by lieut-data** [comment]
@migbot, it looked like I was able to merge the pull request even while there are outstanding linting issues to resolve. Usually, we allow developers to approve a review knowing that it can't be merged without fixing the build issues (vs. having to block the review on a human checking the linting pass). Maybe it's just not a required step in this repository?
